Refs #16859 -- Allowed storing CSRF tokens in sessions.

raphaelm opened the pull request:
I'd love to have some review on this, but I'm fine with postponing the merge after Shai landed his changes to CSRF handling, as those two _will_ get merge conflicts.

charettes commented on django/middleware/csrf.py:
```diff
@@ -105,19 +105,38 @@ def _reject(self, request, reason):
+    def _get_current_token(self, request):
+        try:
+            if settings.CSRF_USE_SESSIONS:
+                return _sanitize_token(request.session[settings.CSRF_COOKIE_NAME])
```
Not sure we want to use `settings.CSRF_COOKIE_NAME` for this. What about `CSRF_SESSION_KEY = '_csrf_token'` just like we do with `LANGUAGE_SESSION_KEY`.

raphaelm replied:
I agree that using settings.CSRF_COOKIE_NAME here is not elegant, but neither is setting it to a hardcoded value and introducing more settings bad as well; but I'm open to discussion ok this.

carljm replied:
I think the session cookie should be independently configurable. Technically there's no reason we'd need two settings, it would be possible to just have `CSRF_SESSION_KEY`, and use it both as the session key and the trigger to use sessions at all (if it's not set, we don't). Not sure if that's actually better than just having two settings.

raphaelm replied:
No, I don't think that is the way to go, as it would be very confusing and inconsistent to read in the settings file.

charettes replied:
I agree that considering setting `CSRF_SESSION_KEY` to a not `None` value as opt-in for enabling session based CSRF protection might be a bit confusing. It's a pattern we've not used in the past and if I remember correctly `LANGUAGE_SESSION_KEY` purpose is only to allow circumventing possible key collisions.

The other downside is that we won't be able to provide a sane default value for this key like we do with `LANGUAGE_SESSION_KEY`. While I think the approach has some merit I believe it's worth introducing two settings here.

shaib replied:
We could rename `CSRF_COOKIE_NAME` as `CSRF_ENTRY_NAME` or `CSRF_ENTRY_KEY`, and then using it as cookie name if we're cookie based and session key if we're session based. This would be an improvement over the current code, but probably not over separate settings.

shaib replied:
Unrelated to the key selection -- why would we need to sanitize a value from the session?

collinanderson commented:
Is this the change that doesn't really provide any extra security but just makes the auditors happy?

raphaelm commented:
@collinanderson In part, yes. 

The suggested change in @apollo13's talk was to sign the CSRF token and this it not what this is doing. However, both putting the token into the session or padding it like @shaib is proposing in #5605 achieves the same goal of auditor happiness.

Storing the CSRF into a session does not add much new security, that is absolutely correct. It more or less only saves you a cookie that has to be transmitted at every HTTP request and saves you a header line in every response.

In theory, when the token is stored in a session, the checks on the ``Referer`` header **should** be no longer necessary (see the long comment [here](https://github.com/django/django/blob/master/django/middleware/csrf.py#L137)), so you could deal better with users not sending Referer headers. However, before we skip this check in this case, we'd like some more experienced security experts to comment on this, otherwise we'll keep it in, better safe than sorry.

shaib commented:
@collinanderson While, as @raphaelm says, this is not what @apollo13 suggested for making auditors happy, it is actually a significantly more valuable contribution in that regard. Keeping CSRF tokens in the session is what most other frameworks do, and so it is what auditors expect; in my experience, almost every security audit of a Django-based site first produces a false positive over Django's unusual CSRF scheme.

@raphaelm no, #5605 is not about auditor happiness but about defending against BREACH.

shaib commented on django/middleware/csrf.py:
```diff
@@ -105,19 +105,38 @@ def _reject(self, request, reason):
+
+    def _set_token(self, request, response):
+        if settings.CSRF_USE_SESSIONS:
+            request.session[settings.CSRF_COOKIE_NAME] = request.META["CSRF_COOKIE"]
```
Assuming we use something other than `CSRF_COOKIE_NAME` as the session key, `"CSRF_COOKIE"` seems like an odd key into `request.META`.

raphaelm replied:
I agree, but I did not want change this, because while it is not a documented API, I'm almost sure that someone out there uses this META entry variable to e.g. supply the key to JavaScrpt.

shaib commented on django/middleware/csrf.py:
```diff
@@ -167,22 +186,31 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
                 if referer.scheme != 'https':
                     return self._reject(request, REASON_INSECURE_REFERER)
 
+                good_referers = []
```
This looks odd. If `CSRF_USE_SESSIONS`, then the `CSRF_COOKIE_DOMAIN` should be irrelevant, and `SESSION_COOKIE_DOMAIN` should be used instead. We shouldn't get more good referrers then we had before.

raphaelm replied:
You are right, I'll revert this.

shaib commented on docs/ref/csrf.txt:
```diff
@@ -132,6 +132,12 @@ The above code could be simplified by using the `JavaScript Cookie library
+    If you activate :setting:`CSRF_USE_SESSIONS`, this method does not work and
+    need to make sure you always include the CSRF token somewhere in your HTML
+    read the token from the DOM with JavaScript.
+
```
Also, the "How it works" section needs to be fixed, and a `.. versionchanged:: 1.10` tag needs to be added.

shaib replied:
On second thought, should we provide the token in a cookie even when sessions are used, to support AJAX?

Or should we provide an AJAX-friendly version of `{% csrf_token %}` which creates a data element rather than a hidden form field?

raphaelm replied:
I'd vote in favor of the second suggestion, but I'd love to hear more opinions on that before rewriting the documentation.

carljm replied:
The first suggestion has the great advantage of seamless backwards-compatibility with all existing Ajax code for Django. Does it have any significant downside?

raphaelm replied:
The possible downside I see right now is that it would only work as expected if ``CSRF_COOKIE_*`` settings are set to the same constraints (domain, path, …) as ``SESSION_COOKIE_*``. So we'd have to ignore completely the ``CSRF_COOKIE_*`` settings in this case, which could lead to confusion.

carljm replied:
Hmm, that is a significant downside. And it does feel a bit weird to have that extraneous cookie set, just for the purposes of supporting Ajax. Putting the token in the markup seems like a more consistent approach when using session-based CSRF. I'm ok with that; it is opt-in of course anyway.

shaib commented on docs/ref/settings.txt:
```diff
@@ -368,6 +368,19 @@ Whether to use a secure cookie for the CSRF cookie. If this is set to ``True``,
+Default: ``False``
+
+Whether to store the CSRF token inside a the user's session instead of a cookie.
+This requires that you are using Django's settings module, see
```
session module

shaib commented on docs/ref/settings.txt:
```diff
@@ -3240,6 +3253,7 @@ Security
   * :setting:`CSRF_COOKIE_NAME`
   * :setting:`CSRF_COOKIE_PATH`
   * :setting:`CSRF_COOKIE_SECURE`
+  * :setting:`CSRF_USE_SESSIONS`
```
Move this to the end of the list (it is sorted alphabetically)

shaib commented on django/conf/global_settings.py:
```diff
@@ -524,6 +524,7 @@
 CSRF_COOKIE_PATH = '/'
 CSRF_COOKIE_SECURE = False
 CSRF_COOKIE_HTTPONLY = False
+CSRF_USE_SESSIONS = False
```
In the DutH sprints, @raphaelm and I considered an option to make `CSRF_USE_SESSIONS` default to "use sessions if sessions are being used in the project". We found it a bit too "magical", so we dropped it.

Since the default project template does use sessions, and is likely to prefer to use sessions for CSRF too, it would be nice to make this default to `True`. Regretfully, this would introduce a backwards-incompatibility with AJAX code.

I suggest, though, that we add `CSRF_USE_SESSIONS = True` to the default new-project template.

raphaelm replied:
If we do so, we'll have to change the documentation on AJAX requests.

carljm replied:
If session-CSRF will require a different approach for AJAX, then that documentation needs to be updated (to show both techniques) regardless of what the startproject template contains. The only different might be a matter of emphasis (which approach is documented first and more prominently). I don't think we should remove the documentation for the existing cookie-based approach.

raphaelm commented:
@shaib I'm absolutely aware that the *intention* of your patch is fixing BREACH, but it also might *help* at making auditors happy as a side effect.

raphaelm commented:
@charettes @carljm @shaib I added an extra CSRF_SESSION_KEY setting, reworked the referer checking and fixed the minor documentation issues. 

I have not yet changed the documentation on AJAX requests or the default project template, as I'm not yet sure whether providing an cookie additionally or suggesting a hidden form field is nicer.

auvipy commented:
could you please complete the remaining works and fix failing tests? @raphaelm

shaib commented:
@auvipy it's me you should be bugging; further work on this is waiting for completion of #5605

timgraham commented:
@raphaelm, would you like to return to this soon now that #5605 is merged?

raphaelm commented:
Yes, I'd be interested in looking into this again! I hope I find time to continue working on it soon.

raphaelm commented:
I just pushed a very basic rebase against current master that makes all existing tests pass. I'll try to give this some more love in the coming days and think about whether new tests or new code paths are necessary.

raphaelm commented:
Okay, I believe the only thing left to do here is add documentation on the AJAX thing. @shaib do you agree or do you see the need for additional tests due to your changes?

auvipy commented on docs/ref/settings.txt:
```diff
@@ -371,6 +371,33 @@ Whether to use a secure cookie for the CSRF cookie. If this is set to ``True``,
+CSRF_USE_SESSIONS
+-----------------
+
+.. versionadded:: 1.10
```
this should be 1.11 now?

auvipy commented on docs/ref/settings.txt:
```diff
@@ -371,6 +371,33 @@ Whether to use a secure cookie for the CSRF cookie. If this is set to ``True``,
+
+CSRF_SESSION_KEY
+----------------
+
```
this too 1.11

raphaelm commented:
Thanks @auvipy, I fixed the version numbers and also added a note to the release notes.

raphaelm commented:
I also added some more AJAX documentation, although I'm unsure about the best JS code to recommend.

**Option 1**
```
{% csrf_token %}
<script type="text/javascript">
// using jQuery
var csrftoken = jQuery("[name=csrfmiddlewaretoken]").val();
</script>
```
Pro: Works the same if the JavaScript part is in an external file instead of inside the template.

**Option 2**
```
{% csrf_token %}
<script type="text/javascript">
// using jQuery
var csrftoken = jQuery('{% csrf_token %}').val();
</script>
```
Pro: Doesn't pollute the DOM
Con: Script needs to be inside a rendered template (and thus doesn't work with Content Security Policy headers)

shaib commented on django/middleware/csrf.py:
```diff
@@ -160,23 +160,47 @@ def _reject(self, request, reason):
-            cookie_token = request.COOKIES[settings.CSRF_COOKIE_NAME]
+            if settings.CSRF_USE_SESSIONS:
+                session_token = request.session[settings.CSRF_SESSION_KEY]
+                csrf_token = _sanitize_token(session_token)
```
Why would you need to sanitize something that's already in your session? Seems a bit late...

raphaelm replied:
Fair point :)

raphaelm replied:
What do you think about users acidentally overriding ``request.session['csrftoken']``? Should we care about this? I think sanitizing (e.g. dropping values of wrong lenth) makes sense in this case. Or raise ``SuspiciousOperation``?

shaib commented on django/middleware/csrf.py:
```diff
@@ -227,15 +251,21 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
                     good_referer = settings.CSRF_COOKIE_DOMAIN
                     server_port = request.get_port()
                     if server_port not in ('443', '80'):
                         good_referer = '%s:%s' % (good_referer, server_port)
```
Here we need
```python
  elif settings.CSRF_COOKIE_DOMAIN is not None and settings.CSRF_USE_SESSIONS:
      raise ImproperlyConfigured("When CSRF uses sessions, use SESSION_COOKIE_DOMAIN instead of CSRF_COOKIE_DOMAIN")
```
This, to make sure people don't just forget such settings, which probably express some security assumptions about their deployment.

raphaelm replied:
Agreed, will change this.

raphaelm replied:
Changed.

shaib commented:
@raphaelm with regard to testing, if it's not too much work, I think there's gain in restructuring the tests as two test classes: The given one, and one which inherits it, and changes the settings to use sessions in its setUpClass and turns them back to cookie in its tearDownClass. Then, by default, you test everything both ways, and you take care of the differences explicitly; whereas now, you're just assuming that what you didn't test specifically with sessions "should work the same as with cookies".

(this is just a suggestion, I don't think it should be a condition for accepting this change)

shaib commented:
With respect to your JS: One con of option 1is that it breaks if there's more than one form on the page, I think.

Either of these two options seem like a workaround rather than a solution, though. I think it would be much better to add an optional argument to the `csrf_token` tag -- something like `{% csrf_token into_js_var="identifier" %}`. Perhaps there should be options to hang the var on some globally-accessible object so that it can be read from non-templated javascript.

raphaelm commented:
> With respect to your JS: One con of option 1is that it breaks if there's more than one form on the page, I think.

I don't believe it would. It would break only if there is a form field called "csrftoken" that isn't a csrf token.

> I think it would be much better to add an optional argument to the csrf_token tag 

What should the template tag output in that case? Everything using <script> variant is still a problem in terms of Content Security Policy.

I'll look into the test structure.

shaib commented:
On Monday 04 July 2016 11:31:09 Raphael Michel wrote:
> > With respect to your JS: One con of option 1is that it breaks if there's
> > more than one form on the page, I think.
> 
> I don't believe it would. It would break only if there is a form field
> called "csrftoken" that isn't a csrf token.
> 

The point is that then there would be more than one field named "csrftoken" for 
the selector to match, and you'll get an iterator instead of a single value.

raphaelm commented:
> The point is that then there would be more than one field named "csrftoken" for
> the selector to match, and you'll get an iterator instead of a single value.

You underestimate the sloppiness of JavaScript. jQuery won't care and just take the first one.

raphaelm commented:
If we want to be explicit, we can always use ``jQuery("[name=csrfmiddlewaretoken]").first().val()``

raphaelm commented:
@shaib I restructured the tests as you suggested.

apollo13 commented on django/conf/global_settings.py:
```diff
@@ -543,6 +543,8 @@ def gettext_noop(s):
 CSRF_COOKIE_SECURE = False
 CSRF_COOKIE_HTTPONLY = False
+CSRF_USE_SESSIONS = False
+CSRF_SESSION_KEY = 'csrftoken'
```
Is there any strong argument to make this configurable and not just have something like https://github.com/django/django/blob/master/django/contrib/auth/__init__.py#L14 ?

raphaelm replied:
The only argument for configurability I see is backwards-compatibility if someone uses a session variable of the name we choose.

apollo13 replied:
Yeah, screw that :D in the worst case make it an attribute on the CSRF class, so a user can override it if we go down the route with an extra class.

apollo13 replied:
Please remove it and put it into a module level variable in the csrf module (and start it with an underscore) -- I really dislike settings, especially when it comes to security ;)

apollo13 commented:
Mhm, technically this PR is tying contrib to core -- are we okay with that?

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -160,23 +161,46 @@ def _reject(self, request, reason):
+
+    def _set_token(self, request, response):
+        if settings.CSRF_USE_SESSIONS:
+            request.session[settings.CSRF_SESSION_KEY] = request.META["CSRF_COOKIE"]
```
Use single quotes to stay consistent with the code above.

apollo13 commented:
On a quick skim over the code looks good to me. A crazy idea (and sorry for bringing it up so late). What would be the downside of introducing a new CSRF middleware class, which lives in `django.contrib.sessions`, inheriting the builtin one and overriding `_get|_set_token`. This would allow us to keep the code simple in the sense that we do not need `if` statements at runtime and would require no additional setting at all (one would just put the mw into the middleware list)

raphaelm commented:
Hey @apollo13, thanks for the review! Using a different subclass that lives in ``django.contrib.sessions`` seems to be another valid option, especially to decouple core from contrib. @shaib What is your opinion on this?

shaib commented:
@raphaelm I think @apollo13 made an interesting suggestion, but there are problematic details. There are the decorators-from-middleware to take care of, and I think even the template tag calls the middleware's `get_token`. This, unless I'm missing something, means that the only way to avoid a setting would involve some mechanism where the active middleware registers itself upon instantiation; I'm not sure that kind of complication is justified, and it may be fragile as well.

shaib commented:
@apollo13 said:
> Mhm, technically this PR is tying contrib to core -- are we okay with that?

I disagree. We only use the session as an attribute on the request; technically, we could use this PR and set `CSRF_USE_SESSIONS=True` and have some user-concocted middleware create a `session` attribute on the request that provides the necessary API, without contrib being involved at all, and it should all work.

shaib commented:
(and sorry, I still don't have time for a proper review of the current code)

apollo13 commented:
@shaib Good point, haven't thought of the decorator… There goes that idea.

raphaelm commented:
Good point, @shaib. Should we raise an ``ImproperlyConfigured`` if a request without ``.session`` hits the middleware with a pointer where to look at?

sjoerdjob commented:
Regarding the JavaScript binding. Would it not make more sense to add a meta tag with the csrf token, and let JS query that? This way you are also not dependent on having to have a form input if there is no form.

apollo13 commented:
I personally often put stuff like that into a data attribute on the body tag, but either way I do not think this should affect this PR request that much, since you already have the same situation if you set CSRF_COOKIE_HTTPONLY

raphaelm commented:
@shaib Let's get this forward :) Shall I just go with @apollo13's last suggestion on the JS part? What else is there to do here?

shaib commented:
@raphaelm a data attribute on the body tag sounds pretty reasonable to me. Other than that -- I promise to look at this next week, but I probably won't be able to get to it earlier.

raphaelm commented:
@shaib Alright, no hurry! Just wanted to make this visible again :)

shaib commented on django/middleware/csrf.py:
```diff
@@ -160,23 +161,46 @@ def _reject(self, request, reason):
-                request.csrf_cookie_needs_reset = True
-            # Use same token next time.
+            response.set_cookie(settings.CSRF_COOKIE_NAME,
+                                request.META["CSRF_COOKIE"],
```
Again single quotes

shaib commented on docs/ref/settings.txt:
```diff
@@ -371,6 +371,33 @@ Whether to use a secure cookie for the CSRF cookie. If this is set to ``True``,
+
+Default: ``False``
+
+Whether to store the CSRF token inside a the user's session instead of
```
in the user's session

shaib commented on docs/ref/settings.txt:
```diff
@@ -371,6 +371,33 @@ Whether to use a secure cookie for the CSRF cookie. If this is set to ``True``,
+Default: ``False``
+
+Whether to store the CSRF token inside a the user's session instead of
+a cookie. This requires that you are using Django's session module, see
```
in a cookie.

shaib replied:
This requires the use of sessions, see :mod:`django.contrib.sessions` for more information

(the point is that while some session implementation is required, it need not be Django's; but we don't really need to get into this nuance, just avoid implying that the session must be Django's)

shaib commented on docs/ref/settings.txt:
```diff
@@ -371,6 +371,33 @@ Whether to use a secure cookie for the CSRF cookie. If this is set to ``True``,
+Default: ``'csrftoken'``
+
+The name of the session key to use for the CSRF authentication token if
+:setting:`CSRF_USE_SESSIONS` is activated. . This can be whatever you want
```
double "full stop" (dot) at end of sentence

shaib replied:
This can be any valid key

(At least with some session backends, keys need to be strings)

shaib commented on docs/ref/settings.txt:
```diff
@@ -3378,6 +3405,8 @@ Security
   * :setting:`CSRF_FAILURE_VIEW`
   * :setting:`CSRF_HEADER_NAME`
   * :setting:`CSRF_TRUSTED_ORIGINS`
+  * :setting:`CSRF_SESSION_KEY`
```
Switch lines 3407 & 3408 to keep settings in alphabetical order

shaib commented on tests/csrf_tests/tests.py:
```diff
@@ -694,3 +689,139 @@ def _set_post(self, post):
+
+    def test_csrf_cookie_age_none(self):
+        pass  # Does not apply to sessions
+
```
All of these empty tests look awkward. I'll try to find a better way to share some tests -- probably along the lines of
```python
class CSRFTestBase(object):
    # all the common parts go here

class CSRFCookieTest(CSRFTestBase, TestCase):
    # Cookie-specific tests and utilities

class CSRFSessionTest(CSRFTestBase, TestCase):
    # Session-specific tests and utilities

shaib commented:
A couple of suggestions seem to be accepted in discussion, but aren't implemented:

- if `CSRF_USE_SESSION` and `request` has no `session` raise `ImproperlyConfigured`
- A sane way to include a token in the HTML where there is no form (for JS; can be deferred)

The only major issue I see with current code is with the tests. I'll try to suggest a fix for it befure DutH.

raphaelm commented:
Hey @shaib thank you very much for the further review. I've added the ``ImproperlyConfigured`` exception and fixed all the styling/documentation issues that you 

I'll try to tackle the testing part later today or tomorrow.

As for the JS part, multiple options have been outlined here. I'm tending towards the ``<script type="text/javascript">var csrftoken = jQuery('{% csrf_token %}').val();</script>`` approach, because while @apollo13's ``<body data-csrf="…">`` approach is a bit cleaner, it has two disadvantages:

* We'd need to add an option to the template tag to output the CSRF token (although that wouldn't be a bad thing to have)
* It requires you to modify your base template that outputs the ``body`` tag, while the other approach works basically anywhere on the page

shaib commented:
I agree with your general argument, but I think `jQuery('{% csrf_token %}').val()` is a bit ugly. I'd feel much better with something like ` <script type="text/javascript">var csrftoken = '{% csrf_token raw %}';</script>` 

I'm going to suggest something about tests shortly

raphaelm commented:
I discussed with @apollo13 at DUTH and we decided that the AJAX method that is currently in the patch (csrf.txt) is the best to go with.

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -160,23 +161,52 @@ def _reject(self, request, reason):
         try:
-            cookie_token = request.COOKIES[settings.CSRF_COOKIE_NAME]
+            if settings.CSRF_USE_SESSIONS:
+                if not getattr(request, 'session', None):
```
Maybe change that into a `try/execpt AttributeError`. It's kinda nitpicky, but given that if you want to use session-based CSRF you will most likely have a session object on the request and then try/except would be faster (And even if not, it seems more natural and shows a nice chained error on python3).

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -160,23 +161,52 @@ def _reject(self, request, reason):
+                        'No session could be found for the request, but CSRF_USE_SETTINGS is '
+                        'turned on.'
+                    )
+                session_token = request.session[settings.CSRF_SESSION_KEY]
```
Merge with the line below since you are not using `session_token` later on. Is it actually needed to sanitize the session token? After all the user cannot change it.

raphaelm replied:
The sanitizing cares about things like "create a new one if it's empty" and things like the upgrade from 1.9-style to 1.10-style tokens (which isn't relevent now that this won't land before 1.11). I'm +/-0 on this.

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -160,23 +161,52 @@ def _reject(self, request, reason):
+            patch_vary_headers(response, ('Cookie',))
+
+    def process_view(self, request, callback, callback_args, callback_kwargs):
+        csrf_token = self._get_current_token(request)
```
Move that below the `csrf_processing_done` -- we do not need to do extra work in that case.

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -227,15 +257,21 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
                         good_referer = '%s:%s' % (good_referer, server_port)
+                elif settings.SESSION_COOKIE_DOMAIN is not None and settings.CSRF_USE_SESSIONS:
+                    good_referer = settings.SESSION_COOKIE_DOMAIN
+                    server_port = request.META['SERVER_PORT']
```
Is there any reason you are not using `request.get_port` like above? I also do not like the duplication, can you set `good_referer` before like:
```
good_referer = settings.SESSION_COOKIE_DOMAIN if settings.USE_CSRF_SESSIONS else settings.CSRF_COOKIE_DOMAIN
```
and then simply check on
```
if good_referer is not None and …
```

shaib replied:
Yes, just be careful not to mess up the `else:` case below

shaib commented:
@raphaelm your latest seems to have removed all the empty-pass tests...

raphaelm commented:
@shaib Oh. Maybe I force-pushed after working on this from different computers :/ Probably I'll then have to look at the tests when I get home from DUTH to avoid duplicate work. Thanks for the notice

apollo13 commented:
@shaib I am not sure how I feel about ``{% csrf_token raw %}`` -- if we are going to change the ``csrf_token`` tag, I'd like to do that a separate pull request afterwards (even if it is more work), to keep the scope of the PR as small as possible.

shaib commented on django/middleware/csrf.py:
```diff
@@ -160,23 +161,52 @@ def _reject(self, request, reason):
+            if settings.CSRF_USE_SESSIONS:
+                if not hasattr(request, 'session'):
+                    raise ImproperlyConfigured(
+                        'CSRF_USE_SETTINGS is enabled, but request.session is not set. Please make '
```
CSRF_USE_SESSIONS

shaib commented:
@raphaelm I managed to find a version with the empty tests in my reflog, will work with that

shaib commented:
https://github.com/raphaelm/django/pull/1

raphaelm commented:
Wow thanks! I'll finish this up first thing tomorrow at the sprints.

shaib commented:
I just had a thought this morning: Do we want to support moves between sessions and cookies? In principle, it's the same token, and it would be really nice. In practice, I suspect two things -- the feature creep @apollo13 mentioned earlier, and obscure security issues which tend to slide exactly in this sort of transition.

apollo13 commented:
@shaib Mhm, good question, it would certainly make upgrading easier?

raphaelm commented:
@shaib

raphaelm commented:
@shaib I also think we should postpone that to later.

raphaelm commented:
@shaib @apollo13 I pushed a new version containing the refactored tests and all of your suggested improvements from yesterday. Only open questions:

* Do we keep the calls to ``_sanitize_token()``? If not, we will have a two code paths for dealing with not-yet-existing tokens, but we save a regex check.

* The upgrade path Shai mentioned. I can sure try to implement that, but expect it to bring us into a hell of conditionals around the ``SESSION_DOMAIN``/``COOKIE_DOMAIN`` thing, if we want to catch any possible edgecase. A best-effort upgrade path that errs on the side of keeping security in all cases but only doing the upgrade in the easy cases could be easier.

apollo13 commented:
I prefer single code paths unless we actually have a performance issue, so I am all for keeping `_sanitize_token()`

raphaelm commented:
Than if it was on me, this'd be now ready for a final review followed by a merge. If you think the upgrade thing is important, I'll have a try at that.

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -160,21 +161,51 @@ def _reject(self, request, reason):
-                request.csrf_cookie_needs_reset = True
-            # Use same token next time.
+        csrf_token = self._get_current_token(request)
+        if csrf_token is not None:
```
This check can be removed since `_get_current_token` takes care of that, no?

raphaelm replied:
Oh, no, this should be finde, I was just wrong in my reasons regarding the sanitization. ``_get_current_token`` actually returns ``None`` if the code does not exist. Taking this into account, I removed the sanitization (because an empty string value shouldn't ever happen).

apollo13 replied:
I disagree, you are already do that in https://github.com/django/django/pull/5600/files#diff-a3be722ce2831a8d11438021d44cedf1R183 -- so this check and setting is redundant.

raphaelm replied:
``_get_current_token`` deliberately returns ``None`` if the cookie or session does not exist (in the except block), so I need to check for this, before I work with it's result. I don't see why it's redundant.

apollo13 replied:
But you are checking if the token `is not None` and then do: `request.META['CSRF_COOKIE'] = csrf_token`. This is the same as https://github.com/django/django/pull/5600/files#diff-a3be722ce2831a8d11438021d44cedf1R183 does. To be honest, I do not really understand your answer either -- can you try explaining a last time in which case this code would get executed and what the result would be?

apollo13 replied:
Or rather, you can delete the assignment from `_get_current_token` and leave the check here.

raphaelm replied:
Ah, I see! Sorry, I got confused by all the comments all over the diff :( So yes, this is redundant. I'm also in favor of keeping the check and removing the assignment from ``_get_current_token`` so ``_get_current_token`` has no unexpected side effects.

shaib commented:
@raphaelm I think the upgrade path can be done separately, and so it should. I suspect a minefield there. We should just open a ticket for it so that we don't forget; I think I'll make it a release blocker to make sure it is considered before 1.11 is out.

I hope I'll have time today for another pass over the code, but I'm not sure.

apollo13 commented:
Yikes, force pushing here makes it really hard to review the changes, please use normal commits if we still change something and we will squash at the end.

raphaelm commented:
@shaib agreed
@apollo13 Will do so

shaib commented:
@apollo13 the github model of pull requests is at fault. gerrit gets this right.

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -160,20 +161,49 @@ def _reject(self, request, reason):
+                    # the cookie needs to be reset.
+                    request.csrf_cookie_needs_reset = True
+            return csrf_token
+        except KeyError:
```
I kinda dislike that `KeyError` spans this much code, but I do not see a nice way around it either. On a minor nitpick I'd also rename `_get_current_token` to just `_get_token` just to be in line with `_set_token`

raphaelm replied:
I removed the setting and I also renamed and rewrote ``_get_token`` to remove the big try/except block. I decided against an underscore in the module-level variable to keep it consistent with the other constants in that file.

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -33,6 +33,7 @@
 CSRF_SECRET_LENGTH = 32
 CSRF_TOKEN_LENGTH = 2 * CSRF_SECRET_LENGTH
 CSRF_ALLOWED_CHARS = string.ascii_letters + string.digits
+CSRF_SESSION_KEY = 'csrftoken'
```
`_csrftoken` please to not collide with user set variables and also use the variable in tests instead of hardcoding

raphaelm replied:
Ah, you meant the underscore in the value not the variable name :D On it!

apollo13 replied:
Hehe :D Thanks, I'll approve now and merge later today after a final testrun

raphaelm replied:
Done. Shall I squash this into one commit now or will that be done while merging?

timgraham commented:
I'll do a docs review today if you can wait a few more hours. :-)

raphaelm commented:
@timgraham After waiting for a year, a few hours will be fine ;) Take your time!

timgraham commented:
Could you submit a separate PR for the test reorganization so we can rebase this and simplify it?

raphaelm commented:
@timgraham I don't think two PRs are a good idea, as the test refactoring is a completely pointless complication of things if you don't view it in the context of this PR. However, I could imagine refactoring the change into two or more *commits* to make review easier, if that helps.

timgraham commented:
Maybe a second commit that moves the existing tests would help. If you don't think it makes sense to merge those separate commits, then don't bother.

Could you add an explanation of why someone would want to use this setting? (in the settings docs and/or ref/csrf.txt)

The cosmetic edits: http://dpaste.com/3QG55D1

raphaelm commented:
@timgraham I reworked the patch with your suggestions. I started to try splitting it into multiple PRs/commits, but stopped as it doesn't seem to make a lot of sense. The seperate commit ("test refactoring") would be the exact same thing we have in here except for the ``CsrfViewMiddlewareUseSessionsTest`` class and the second commit would be adding the ``CsrfViewMiddlewareUseSessionsTest`` class. I don't think adding the class in a different commit would make the diff significantly easier to read.

@apollo13 I squashed these latest changes with the last commit (not with all of them), sorry for that (happened after some local git confusion and some reflog fun)

timgraham commented on django/middleware/csrf.py:
```diff
@@ -160,20 +162,50 @@ def _reject(self, request, reason):
+                domain=settings.CSRF_COOKIE_DOMAIN,
+                path=settings.CSRF_COOKIE_PATH,
+                secure=settings.CSRF_COOKIE_SECURE,
+                httponly=settings.CSRF_COOKIE_HTTPONLY
```
add trailing comma

timgraham commented on django/middleware/csrf.py:
```diff
@@ -226,16 +258,22 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
+                    if settings.CSRF_USE_SESSIONS
+                    else settings.CSRF_COOKIE_DOMAIN
+                )
+
```
I suggested to remove the blank line because the ) is already providing space.

timgraham commented on docs/ref/csrf.txt:
```diff
@@ -64,6 +64,16 @@ XMLHttpRequest, set a custom ``X-CSRFToken`` header to the value of the CSRF
+set if you've enabled CSRF protection for your views as outlined above.
+
+
 As a first step, you must get the CSRF token itself. The recommended source for
```
My diff didn't get applied correctly here. This paragraph was removed.

timgraham commented on docs/releases/1.11.txt:
```diff
@@ -220,6 +220,9 @@ Cache
+* Added the :setting:`CSRF_USE_SESSIONS` setting to allow storing the CSRF
+  token in the user's session rather than in a cookie.
+
 * ...
```
Remove this placeholder.

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -160,20 +162,50 @@ def _reject(self, request, reason):
+            patch_vary_headers(response, ('Cookie',))
+
+    def process_view(self, request, callback, callback_args, callback_kwargs):
+        if getattr(request, 'csrf_processing_done', False):
```
`getattr(request, 'csrf_processing_done')` would suffice since `None` is not true.

apollo13 replied:
Oh sorry, I did mix it up with `dict.get` -- /me goes into the shame corner

apollo13 commented:
Ok, from a technical perspective I have no more issues with that PR, @timgraham please merge when the doc fixes are taken care of.

raphaelm commented:
> Did you see my comment asking to document the use case for enabling CSRF_USE_SESSIONS? I don't think it was addressed.

Sorry, I completely forgot about that, my bad. I addressed that now, as well as all the other comments you added yesterday.

To ease reviewing, I did not squash my commits yet. Please tell me if I should do so or just squash them yourself, as you prefer. Thank you all for the extensive and helpful review, @timgraham, @shaib, @apollo13 and the others.

timgraham commented on tests/csrf_tests/tests.py:
```diff
@@ -652,45 +602,214 @@ def test_csrf_cookie_age_none(self):
+        req.session[CSRF_SESSION_KEY] = self._csrf_id_cookie
+        return req
+
+    def setUp(self):
```
Did you try doing this in `setUpClass` or using `@override_settings` on the class?

raphaelm replied:
``@override_settings`` apparently works, I don't know why I didn't use it in the first place

timgraham commented:
Could you add a test for the new `ImproperlyConfigured`? Few more cosmetic stuff: http://dpaste.com/0XHN81P Feel free to squash when updating. Thanks!

raphaelm commented:
Cosmetics fixed, squashed and rebased against current master :)

timgraham commented:
merged in ddf169cdaca91e92dd5bfe6796bb6f38369ecb68, thanks!

timgraham commented:
@raphaelm, seems like the security check for `CSRF_COOKIE_SECURE` needs to be disabled when using `CSRF_USE_SESSIONS`. Can you do it? (Same for `CSRF_COOKIE_HTTPONLY` except that it'lll be removed in https://github.com/django/django/pull/7700).

raphaelm commented:
@timgraham Sure, will do
